jcl/src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -810,7 +810,7 @@ public final boolean isDaemon() {
 public boolean isInterrupted() {
 	synchronized(lock) {
 		/*[IF Java14]*/
-		if (!isAlive()) {
+		if (threadRef == NO_REF) {
```

Comment:
I don't think this does anything - since you already own the lock, calling `isAlive` is identical to this code, which might make this solution incorrect.

Reply:
I suspect this will work by chance - the fact that we only release VM access on certain operations (like method enter) might save us.

Reply:
We should add a comment that we are testing for isAlive without calling the method.

Reply:
If @gacholio isn't willing to accept this change with the addition of the comments, then the natives need to be modified to return and set deadInterrupt instead.

Reply:
Manually inlining the `isAlive` call here is very brittle.  There should be a detailed comment referencing `isAlive` & describing why we've manually inlined it here - namely to avoid anything that might check vmaccess

Reply:
I'm not familiar with this code at all so forgive me if not a relevant suggestion, but wouldn't you also want a comment at `isAlive` to point at this code location, in case `isAlive` ever updated with new semantics that would be also needed at this code location?

Reply:
@pshipton It's not a matter of me "not accepting" the change - it obviously does nothing other than avoiding a method enter/async check.

Reply:
The VM will async check on any taken branch, which in this case means the check will occur if the thread is alive (the expected case). Even if this weren't the case, building this kind of knowledge into the java code is a terrible idea.

Reply:
ok, while the current code works and avoids the problems, it seems the consensus is to put the deadInterrupt code into the natives instead @ChengJin01

Reply:
If you really believe this fixes anything, go ahead and commit it as-is. The issue suspending a thread while it holds one of the thread locks is general - it will show up in an number of other methods. This essentially can't be fixed without moving all the locking code into the natives, which is a large amount of work.

Reply:
Getting rid of the async check from calling isAlive() does appear to fix the issues caused by https://github.com/eclipse/openj9/pull/9341. This PR also removes https://github.com/eclipse/openj9/pull/9907 which introduced another problem. However as you say, it's fragile to do this in the java code. It won't be hard to move the deadInterrupt boolean logic into the natives so it's not so fragile. We can defer moving all the locking code, as this is a long standing problem that nobody is requesting be fixed atm.